src/headers.jl

```diff
@@ -0,0 +1,246 @@
+# See https://httpwg.github.io/specs/rfc7230.html#rule.token.separators
+const validHeaderBytes = IntSet(Int(c) for c in
+   raw"!#$%&'*+-.0123456789ABCDEFGHIJKLMNOPQRSTUWVXYZ^_`abcdefghijklmnopqrstuvwxyz|~"
+)
+
+validHeaderChar(c::Char) = isascii(c) && UInt8(c) in validHeaderBytes
+
+function canonicalHeaderKey(s::String)
+    # fast path for valid header keys
+    upper = true
+    for c in s
+        if !validHeaderChar(c)
+            return s
+        end
+        if upper && 'a' <= c <= 'z'
+            return _chk(s)
+        end
+        if !upper && 'A' <= c <= 'Z'
+            return _chk(s)
+        end
+
+        upper = c == '-'
+    end
+
+    return s
+end
+
+const toUpper = 'A' - 'a'
```

Comment:
this is only used in `_chk`, so we can avoid the global and just put it at the top of the function definition.